### PR TITLE
Allow ROOT_INCLUDE to specify standard library include paths

### PR DESCRIPTION
**Pull request by Y--**
This way we do not need g++ at runtime

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/CIFactory.cpp [hunk 1]
```diff
@@ -426,12 +435,19 @@ namespace {
         }
       }
 #else // _MSC_VER
+      const char* RootInclude = std::getenv("ROOT_INCLUDE");
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: We try to layer ROOT on top of cling, as cleanly as possible. Thus cling should use ROOT_INCLUDE. Can we maybe adjust the interpreter construction instead, passing `-isystem` and `-nostdinc++`? Then cling could skip the stdinc++ determination, reacting on `-nostdinc++` being passed as an argument.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp [hunk 2]
```diff
@@ -91,6 +91,19 @@ std::unique_ptr<TargetMachine>
   CodeModel::Model CMModel = CodeModel::JITDefault;
   CodeGenOpt::Level OptLevel = CodeGenOpt::Less;
 
+  const char* OptLevelEnvConfig = std::getenv("ROOT_OPTIMIZATION_LEVEL");
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Can we react on -O0 etc passed to the cling constructor, and pass the -On flag in when constructing the interpreter in core/meta/src/TCling.cxx?

**Comment by karies** [comment]
Thank you for your PR! Nice ideas, let's get this in! Please let me know if one of my comments wasn't precise enough or misunderstood your patch.

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp (hunk 2 above)
Comment (Y--) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: That sounds much better indeed.
I need to look deeper into the code to achieve that though.

Just to be clear, you're suggesting to pass the `-Ox` flags to the `args` of
`cling::Interpreter`'s constructor, and then using it here?

File: interpreter/cling/lib/Interpreter/CIFactory.cpp (hunk 1 above)
Comment (Y--) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: That would be great and makes much more sense indeed.
Is it something that is already implemented or does it need to be?

**Comment by Y--** [comment]
Thanks for your quick review!
I'll try to work on your suggestions asap so you can integrate them.
Also, do you want me to do the same PR on Cling project?

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp (hunk 2 above)
Comment (karies) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Correct. Thanks!

File: interpreter/cling/lib/Interpreter/CIFactory.cpp (hunk 1 above)
Comment (karies) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: Handling of -isystem should exist; skipping the gcc invocation if -nostdinc++ is passed or the ROOT part of the compiler includes do not exist.

**Comment by karies** [comment]
@Y-- - no need for an extra PR for cling!

File: interpreter/cling/lib/Interpreter/CIFactory.cpp (hunk 1 above)
Comment (Y--) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: It works properly indeed.
Here is the new version : https://github.com/Y--/root/commit/39736e1e362a0871d3a5a52b39305a67220ccd1a

File: interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp (hunk 2 above)
Comment (Y--) [inline interpreter/cling/lib/Interpreter/IncrementalExecutor.cpp]: Implemented here : https://github.com/Y--/root/commit/97a6d48dc1cd40ab414d76cd6bbfe48343ba144d and there https://github.com/Y--/root/commit/5f8c40035a4f9305a53384a35419d3f4d508021b

**Comment by karies** [comment]
@Y-- Indeed, last of -O3 -O0 wins.

File: interpreter/cling/lib/Interpreter/CIFactory.cpp [hunk 3]
```diff
@@ -41,6 +41,7 @@
 #include <cstdio>
 
 #include <memory>
+#include <sstream>
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: Where's this needed?

Comment (Y--) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: Oups, I forgot to remove that. It's not used indeed.
Removed here : https://github.com/Y--/root/commit/83e6fc71efef78ce45d7129550e6c995afbac3a1

**Comment by karies** [comment]
Excellent! Merged as 10cf52d..73b3a1d into master. (I have squashed some of your commits.)

Thank you for your contributions!

**Comment by Y--** [comment]
Awesome! Thank you so much :-)

**Pull request closed without merging**